File: src/maluuba/speech/pronunciation/phone.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "maluuba/speech/pronunciation/impl.hpp"
#include "maluuba/speech/pronunciation.hpp"
#include "maluuba/debug.hpp"
#include <cstdint>

namespace maluuba
{
namespace speech
{
  using namespace internal;

  namespace
  {
    void
    check_consonant(const Phone& phone)
    {
      check_logic(phone.type() == PhoneType::CONSONANT, "This phone is not a consonant.");
    }

    void
    check_vowel(const Phone& phone)
    {
      check_logic(phone.type() == PhoneType::VOWEL, "This phone is not a vowel.");
    }
  }

  Phone::Phone(PhoneType type)
    : Phone{phone_encode(type, type_start)}
  { }

  Phone::Phone(std::uint16_t repr)
    : m_repr{repr}
  { }

  PhoneType
  Phone::type() const
  {
    return phone_decode<PhoneType>(m_repr, type_start, type_end);
  }

  Phonation
  Phone::phonation() const
  {
    return phone_decode<Phonation>(m_repr, phonation_start, phonation_end);
  }

  void
  Phone::phonation(Phonation phonation)
  {
    m_repr = phone_encode(m_repr, phonation, phonation_start, phonation_end);
  }

  PlaceOfArticulation
  Phone::place() const
  {
    check_consonant(*this);
    return phone_decode<PlaceOfArticulation>(m_repr, place_start, place_end);
  }

  void
  Phone::place(PlaceOfArticulation place)
  {
    check_consonant(*this);
    m_repr = phone_encode(m_repr, place, place_start, place_end);
  }

  MannerOfArticulation
  Phone::manner() const
  {
    check_consonant(*this);
    return phone_decode<MannerOfArticulation>(m_repr, manner_start, manner_end);
  }

  void
  Phone::manner(MannerOfArticulation manner)
  {
    check_consonant(*this);
    m_repr = phone_encode(m_repr, manner, manner_start, manner_end);
  }

  VowelHeight
  Phone::height() const
  {
    check_vowel(*this);
    return phone_decode<VowelHeight>(m_repr, height_start, height_end);
  }

  void
  Phone::height(VowelHeight height)
  {
    check_vowel(*this);
    m_repr = phone_encode(m_repr, height, height_start, height_end);
  }

  VowelBackness
  Phone::backness() const
  {
    check_vowel(*this);
    return phone_decode<VowelBackness>(m_repr, backness_start, backness_end);
  }

  void
  Phone::backness(VowelBackness backness)
  {
    check_vowel(*this);
    m_repr = phone_encode(m_repr, backness, backness_start, backness_end);
  }

  VowelRoundedness
  Phone::roundedness() const
  {
    check_vowel(*this);
    return phone_decode<VowelRoundedness>(m_repr, roundedness_start, roundedness_end);
  }

  void
  Phone::roundedness(VowelRoundedness roundedness)
  {
    check_vowel(*this);
    m_repr = phone_encode(m_repr, roundedness, roundedness_start, roundedness_end);
  }

  bool
  Phone::is_rhotic() const
  {
    check_vowel(*this);
    return phone_decode<bool>(m_repr, rhotic_start, rhotic_end);
  }

  void
  Phone::rhotic(bool rhotic)
  {
    check_vowel(*this);
    m_repr = phone_encode(m_repr, rhotic, rhotic_start, rhotic_end);
  }

  bool
  Phone::is_syllabic() const
  {
    return phone_decode<bool>(m_repr, syllabic_start, syllabic_end);
  }

  void
  Phone::syllabic(bool syllabic)
  {
    m_repr = phone_encode(m_repr, syllabic, syllabic_start, syllabic_end);
  }
}
}

File: src/cs/Microsoft.PhoneticMatching/Phone.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.PhoneticMatching
{
    /// <summary>
    /// Phone type (consonant or vowel)
    /// </summary>
    public enum PhoneType
    {
        Consonant,
        Vowel
    }

    /// <summary>
    /// Phonation (voice intensity)
    /// </summary>
    public enum Phonation
    {
        Voiceless,
        Breathy,
        Slack,
        Modal,
        Stiff,
        Creaky,
        GlottalClosure,
    }

    /// <summary>
    /// Place of articulation for consonants.
    /// </summary>
    public enum PlaceOfArticulation
    {
        Bilabial,
        Labiodental,
        Dental,
        Alveolar,
        PalatoAlveolar,
        Retroflex,
        AlveoloPalatal,
        LabialPalatal,
        Palatal,
        PalatalVelar,
        LabialVelar,
        Velar,
        Uvular,
        Pharyngeal,
        Epiglottal,
        Glottal,
    }

    /// <summary>
    /// Manner of articulation for consonants.
    /// </summary>
    public enum MannerOfArticulation
    {
        Nasal,
        Plosive,
        SibilantFricative,
        NonSibilantFricative,
        Approximant,
        Flap,
        Trill,
        LateralFricative,
        LateralApproximant,
        LateralFlap,
        Click,
        Implosive,
        Ejective
    }

    /// <summary>
    /// Vowel height.
    /// </summary>
    public enum VowelHeight
    {
        Close,
        NearClose,
        CloseMid,
        Mid,
        OpenMid,
        NearOpen,
        Open
    }

    /// <summary>
    /// Horizontal vowel position.
    /// </summary>
    public enum VowelBackness
    {
        Front,
        NearFront,
        Central,
        NearBack,
        Back
    }

    /// <summary>
    /// Vowel roundedness.
    /// </summary>
    public enum VowelRoundedness
    {
        Unrounded,
        LessRounded,
        Rounded,
        MoreRounded
    }

    /// <summary>
    /// A phone is a unit of speech sound.
    /// </summary>
    public struct Phone
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="Phone"/> struct.
        /// </summary>
        /// <param name="type">the phone type (consonant or vowel)</param>
        /// <param name="phonation">the phonation (voice intensity)</param>
        /// <param name="place">the place of articulation for consonants.</param>
        /// <param name="manner">the manner of articulation for consonants.</param>
        /// <param name="height">the vowel height.</param>
        /// <param name="backness">the horizontal vowel position.</param>
        /// <param name="roundedness">the vowel roundedness.</param>
        /// <param name="isRhotic">a value indicating whether the phone is rhotic.</param>
        /// <param name="isSyllabic">a value indicating whether the phone is syllabic.</param>
        public Phone(
            PhoneType type, 
            Phonation phonation, 
            PlaceOfArticulation place, 
            MannerOfArticulation manner, 
            VowelHeight height, 
            VowelBackness backness,
            VowelRoundedness roundedness,
            bool isRhotic,
            bool isSyllabic)
        {
            this.Type = type;
            this.Phonation = phonation;
            this.IsSyllabic = isSyllabic;

            if (type == PhoneType.Vowel)
            {
                this.Height = height;
                this.Backness = backness;
                this.Roundedness = roundedness;
                this.IsRhotic = isRhotic;
                this.Place = null;
                this.Manner = null;
            }
            else
            {
                this.Height = null;
                this.Backness = null;
                this.Roundedness = null;
                this.IsRhotic = null;
                this.Place = place;
                this.Manner = manner;
            }
        }

        /// <summary>
        /// Gets the phone type (consonant or vowel)
        /// </summary>
        public PhoneType Type { get; private set; }

        /// <summary>
        /// Gets the phonation (voice intensity)
        /// </summary>
        public Phonation Phonation { get; private set; }

        /// <summary>
        /// Gets the place of articulation for consonants.
        /// </summary>
        public PlaceOfArticulation? Place { get; private set; }

        /// <summary>
        /// Gets the manner of articulation for consonants.
        /// </summary>
        public MannerOfArticulation? Manner { get; private set; }

        /// <summary>
        /// Gets the vowel height.
        /// </summary>
        public VowelHeight? Height { get; private set; }

        /// <summary>
        /// Gets the horizontal vowel position.
        /// </summary>
        public VowelBackness? Backness { get; private set; }

        /// <summary>
        /// Gets the vowel roundedness.
        /// </summary>
        public VowelRoundedness? Roundedness { get; private set; }

        /// <summary>
        /// Gets a value indicating whether the phone is rhotic.
        /// </summary>
        public bool? IsRhotic { get; private set; }

        /// <summary>
        /// Gets a value indicating whether the phone is syllabic.
        /// </summary>
        public bool IsSyllabic { get; private set; }
    }
}
